### PR TITLE
JAV-501 Public key authentication between consumer and provider

### PR DESCRIPTION
**Pull request by coolhongluo**

### CODE REVIEW

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  public static RSAKeyPairEntry getRSAKeyPair() {
+    try {
+      KeyPairGenerator keyGenerator = KeyPairGenerator.getInstance(RSA_ALG);
+      keyGenerator.initialize(KEY_SIZE, new SecureRandom());
+      KeyPair keyPair = keyGenerator.generateKeyPair();
+      PublicKey pubKey = keyPair.getPublic();
+      PrivateKey privKey = keyPair.getPrivate();
+      return new RSAKeyPairEntry(privKey, pubKey, encoder.encodeToString(pubKey.getEncoded()));
+    } catch (NoSuchAlgorithmException e) {
+      throw new Error(e);
```
Comment (liubao68) [inline foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java]: Avoid using Error, when you need throw some unexpected exception, using RuntimeException, e.g. IllegalStageException or IllgaleArgumentException

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+  private final static String SIGN_ALG = "SHA256withRSA";
+  
+  private final static int KEY_SIZE = 2048;
+
+  private static Base64.Encoder encoder = Base64.getEncoder();
+
+  private static Base64.Decoder decoder = Base64.getDecoder();
+
+  public static RSAKeyPairEntry getRSAKeyPair() {
```
Comment (liubao68) [inline foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java]: It's better to use generateRSAKeyPair

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/provider/RSAProviderTokenManager.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+import io.servicecomb.authentication.RSAAuthenticationToken;
+import io.servicecomb.foundation.common.utils.RSAUtils;
+import io.servicecomb.serviceregistry.api.Const;
+import io.servicecomb.serviceregistry.api.registry.MicroserviceInstance;
+import io.servicecomb.serviceregistry.cache.MicroserviceInstanceCache;
+
+public class RSAProviderTokenManager  {
+
+  private static Logger logger = LoggerFactory.getLogger(RSAProviderTokenManager.class);
```
Comment (liubao68) [inline handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/provider/RSAProviderTokenManager.java]: private static final Logger LOGGER

Comment (coolhongluo) [reply handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/provider/RSAProviderTokenManager.java]: accept

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java (hunk 2 above)
Comment (coolhongluo) [reply foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java]: accept

**Comment by WillemJiang** [comment]
@coolhongluo  Please resolve the conflict and fix the small issues which liubao already pointed out.

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAKeyPairEntry.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package io.servicecomb.foundation.common.utils;
```
Comment (WillemJiang) [inline foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAKeyPairEntry.java]: Please add the license header.

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
+package io.servicecomb.foundation.common.utils;
```
Comment (WillemJiang) [inline foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java]: License header.

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java (hunk 1 above)
Comment (WillemJiang) [reply foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java]: It's better to write a log for it.

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+      PrivateKey privKey = keyPair.getPrivate();
+      return new RSAKeyPairEntry(privKey, pubKey, encoder.encodeToString(pubKey.getEncoded()));
+    } catch (NoSuchAlgorithmException e) {
+      throw new Error(e);
+    }
+  }
+
+  public static String sign(String content, PrivateKey privateKey)
+      throws NoSuchAlgorithmException, InvalidKeySpecException, SignatureException, InvalidKeyException {
+    Signature signature = Signature.getInstance(SIGN_ALG);
```
Comment (wujimin) [inline foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java]: instance can be reused?
i had asked this question in last PR.

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    signature.update(content.getBytes());
+    byte[] signByte = signature.sign();
+    return encoder.encodeToString(signByte);
+  }
+
+  public static boolean verify(String publicKey, String sign, String content)
+      throws NoSuchAlgorithmException, InvalidKeySpecException, InvalidKeyException, SignatureException {
+    byte[] bytes = decoder.decode(publicKey);
+    X509EncodedKeySpec keySpec = new X509EncodedKeySpec(bytes);
+    KeyFactory kf = KeyFactory.getInstance(RSA_ALG);
```
Comment (wujimin) [inline foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java]: instance can be reused?
i had asked this question in last PR.

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+import org.apache.commons.lang3.RandomStringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.servicecomb.authentication.RSAAuthenticationToken;
+import io.servicecomb.foundation.common.utils.RSAUtils;
+import io.servicecomb.foundation.token.RSAKeypair4Auth;
+import io.servicecomb.serviceregistry.RegistryUtils;
+
+public class RSACoumserTokenManager {
```
Comment (wujimin) [inline handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java]: consumer?
coumser?

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+  private static final Logger logger = LoggerFactory.getLogger(RSACoumserTokenManager.class);
+
+  private ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+
+  private RSAAuthenticationToken token;
+
+  public String getToken() {
+    readWriteLock.readLock().lock();
+    if (isvalid(token)) {
```
Comment (wujimin) [inline handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java]: valid in this time, but expired in next nanoSecond......
so producer must think it's expired.

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java (hunk 1 above)
Comment (coolhongluo) [reply foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java]: accept

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+      String tokenStr = token.format();
+      readWriteLock.readLock().unlock();
+      return tokenStr;
+    } else {
+      readWriteLock.readLock().unlock();
+      return createToken();
```
Comment (wujimin) [inline handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java]: it seems you are writing a double check logic.
if true, then your code is too complex.

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+  private static final Logger logger = LoggerFactory.getLogger(RSACoumserTokenManager.class);
+
+  private ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+
+  private RSAAuthenticationToken token;
+
+  public String getToken() {
+    readWriteLock.readLock().lock();
+    if (isvalid(token)) {
+      String tokenStr = token.format();
```
Comment (wujimin) [inline handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java]: format result never changed, cache it.

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+   * the TTL of Token is  24 hours
+   * client token will expired 15 minutes early 
+   */
+  public boolean isvalid(RSAAuthenticationToken token) {
+    if (null == token) {
+      return false;
+    }
+    long generateTime = token.getGenerateTime();
+    Date expiredDate = new Date(generateTime + RSAAuthenticationToken.TOKEN_ACTIVE_TIME - 15 * 60 * 1000);
+    Date now = new Date();
```
Comment (wujimin) [inline handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java]: same to java.lang.System.currentTimeMillis()
why need to new Date()?

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    if (now.before(expiredDate)) {
```
Comment (wujimin) [inline handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java]: return now.before(expiredDate);

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/provider/RSAProviderTokenManager.java [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        return false;
+      }
+      if (tokenExprired(rsaToken)) {
+        logger.error("token is expired");
+        return false;
+      }
+      if (validatedToken.contains(rsaToken)) {
+        logger.info("found vaildate token in vaildate pool");
+        return true;
+      } else {
```
Comment (wujimin) [inline handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/provider/RSAProviderTokenManager.java]: already return, no need "else"

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java (hunk 6 above)
Comment (coolhongluo) [reply foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java]: 通过查看源代码和使用步骤来看，Signature 是有状态的。

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java (hunk 7 above)
Comment (coolhongluo) [reply foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java]: accept KeyFactory instance can reused

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java (hunk 8 above)
Comment (coolhongluo) [reply handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java]: accept

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/provider/RSAProviderTokenManager.java (hunk 14 above)
Comment (coolhongluo) [reply handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/provider/RSAProviderTokenManager.java]: accept

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java (hunk 6 above)
Comment (wujimin) [reply foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java]: "This signature object is reset to its initial state (the state it was in after a call to one of the initSign methods) and can be reused to generate further signatures with the same private key."

copy from java doc.

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAKeyPairEntry.java (hunk 4 above)
Comment (coolhongluo) [reply foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAKeyPairEntry.java]: accept

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java (hunk 5 above)
Comment (coolhongluo) [reply foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java]: accept

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java (hunk 9 above)
Comment (coolhongluo) [reply handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java]: 对于这种场景，误差允许的，token过时策略并不是绝对的过时，主要是防止token泄露后 永远有效的漏洞

**Pull request closed without merging**